File: app/domains/acts/formatters/table_title.py
"""Единое правило оформления подписи таблицы — зеркало фронта.

Правило владельца, действующее во ВСЕХ контурах (редактор, превью, диалог
версий, DOCX/MD/TXT):

- пресетная (автосозданная) таблица разделов 1–4 — обычным начертанием,
  ПОДЧЁРКНУТАЯ;
- раздел 5, любые таблицы — обычным начертанием, БЕЗ подчёркивания;
- пользовательская таблица — без эффектов;
- таблица внутри блока нарушения — без эффектов (подписи у неё нет вовсе).

Жирного начертания у подписи нет ни в одном контуре.

Фронтовое зеркало — ``static/js/constructor/table/table-title.js``
(``UNDERLINED_TITLE_SECTIONS`` / ``tableTitleUnderlined``); импорт из Python
невозможен, поэтому синхронность сторожит
``tests/domains/acts/formatters/test_table_title_rule.py``: одна матрица
случаев прогоняется через обе реализации.

Раздел узла приходит из ``WalkContext.root_section_id`` (его проставляет
tree_walker), поэтому правило одинаково доступно всем трём форматтерам.
"""
from __future__ import annotations

from typing import Any, Mapping

from app.domains.acts.block_types import NODE_TYPE_TABLE

# Разделы, у которых пресетная таблица получает подчёркнутую подпись.
# Правило владельца сформулировано ПО РАЗДЕЛАМ: 1–4 — подчёркиваем, 5 — нет.
UNDERLINED_TITLE_SECTIONS: frozenset[str] = frozenset({"1", "2", "3", "4"})

# Подвид обычной таблицы (зеркало KIND_REGULAR фронта / TABLE_KINDS схемы).
_KIND_REGULAR = "regular"


def is_preset_table(node: Mapping[str, Any] | None) -> bool:
    """Пресетная (автоматически созданная) таблица.

    Признака «создана автоматически» в модели нет, дискриминатор — защищённая
    обычная таблица: у пользовательских ``protected`` False, у спецтаблиц
    (метрики/риски) ``kind`` отличен от 'regular'.
    """
    if not node:
        return False
    return (
        node.get("type") == NODE_TYPE_TABLE
        and node.get("protected") is True
        and (node.get("kind") or _KIND_REGULAR) == _KIND_REGULAR
    )


def table_title_underlined(
    node: Mapping[str, Any] | None, root_section_id: str | None,
) -> bool:
    """Подчёркивается ли подпись таблицы (единое правило всех контуров).

    Args:
        node: Узел-таблица дерева акта.
        root_section_id: ID раздела верхнего уровня, под которым лежит узел
            (``WalkContext.root_section_id``); None — раздел неизвестен.

    Returns:
        True, если подпись подчёркивается.
    """
    if root_section_id is None:
        return False
    return is_preset_table(node) and str(root_section_id) in UNDERLINED_TITLE_SECTIONS

File: static/js/constructor/table/table-title.js
/**
 * Единый предикат и текст заголовка таблицы — общий для DOM-рендерера
 * (items-renderer) и превью (preview). Без него условия показа расходились:
 * рендерер/превью скрывали заголовок при customLabel==='' (дефолт защищённых
 * таблиц из act_crud_service), хотя DOCX-экспорт в этом случае показывает
 * label (см. app/domains/acts/formatters/docx/formatter.py::_add_table_title:
 * `title = customLabel or label; if not title: return`).
 *
 * Эталон — DOCX: заголовок показывается, если есть непустой customLabel ИЛИ
 * label. Фронт дополнительно поддерживает автонумерацию (number) как
 * запасной текст, поэтому в предикат включён и number — это надмножество
 * DOCX, не сужающее показ.
 *
 * Здесь же — единое правило ОФОРМЛЕНИЯ подписи (tableTitleUnderlined): до него
 * оформление было размазано по контурам (инлайн-стили редактора, CSS превью,
 * bold в DOCX, дефисы в TXT) и разъехалось. Питоновское зеркало правила —
 * app/domains/acts/formatters/table_title.py, страж синхронизации —
 * tests/domains/acts/formatters/test_table_title_rule.py.
 *
 * Модуль БЕЗ DOM-зависимостей (импортируется и из тестов напрямую).
 */
import { getTableKind, KIND_REGULAR } from './table-kind.js';

/** Тип узла-таблицы (зеркало AppConfig.nodeTypes.TABLE, как в table-kind.js). */
const NODE_TYPE_TABLE = 'table';

/**
 * Разделы, у которых пресетная таблица получает подчёркнутую подпись.
 * Правило владельца сформулировано ПО РАЗДЕЛАМ: 1–4 — подчёркиваем, 5 — нет.
 * @type {readonly string[]}
 */
export const UNDERLINED_TITLE_SECTIONS = Object.freeze(['1', '2', '3', '4']);

/**
 * Текст заголовка таблицы: пользовательская метка → автонумерация → label.
 * @param {Object} node Узел-таблица дерева.
 * @returns {string} Текст заголовка ('' если показывать нечего).
 */
export function tableTitleText(node) {
    return node.customLabel || node.number || node.label || '';
}

/**
 * Показывать ли заголовок таблицы. Эталон — DOCX: непустой customLabel/label
 * (+ number как фронтовый fallback). customLabel==='' больше не скрывает
 * заголовок, если есть number/label.
 * @param {Object} node Узел-таблица дерева.
 * @returns {boolean} true если заголовок нужно отрисовать.
 */
export function shouldShowTableTitle(node) {
    return tableTitleText(node) !== '';
}

/**
 * Пресетная (автоматически созданная) таблица: признака «создана автоматически»
 * в модели нет, дискриминатор — защищённая обычная таблица. У пользовательских
 * protected=false, у спецтаблиц (метрики/риски) kind !== 'regular'.
 * @param {Object} node Узел-таблица дерева.
 * @returns {boolean} true если таблица пресетная.
 */
export function isPresetTable(node) {
    return !!node
        && node.type === NODE_TYPE_TABLE
        && node.protected === true
        && getTableKind(node) === KIND_REGULAR;
}

/**
 * Подчёркивать ли подпись таблицы — единое правило всех контуров.
 * Подчёркивается только пресетная таблица разделов 1–4; раздел 5, любые
 * пользовательские и спецтаблицы — без эффектов (жирного нет нигде).
 * @param {Object} node Узел-таблица дерева.
 * @param {string|null} rootSectionId ID раздела верхнего уровня, под которым лежит узел.
 * @returns {boolean} true если подпись подчёркивается.
 */
export function tableTitleUnderlined(node, rootSectionId) {
    if (rootSectionId === null || rootSectionId === undefined) return false;
    return isPresetTable(node)
        && UNDERLINED_TITLE_SECTIONS.includes(String(rootSectionId));
}

// Дублируем в window ради inline-скриптов в шаблонах (см. CLAUDE.md).
// Guard: модуль также импортируется в node:test, где window отсутствует.
if (typeof window !== 'undefined') {
    window.tableTitleText = tableTitleText;
    window.shouldShowTableTitle = shouldShowTableTitle;
    window.tableTitleUnderlined = tableTitleUnderlined;
}
